### PR TITLE
[SPARK-22975][SS] MetricsReporter should not throw exception when there was no progress reported

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

`MetricsReporter ` assumes that there has been some progress for the query, ie. `lastProgress` is not null. If this is not true, as it might happen in particular conditions, a `NullPointerException` can be thrown.

The PR checks whether there is a `lastProgress` and if this is not true, it returns a default value for the metrics.

## How was this patch tested?

added UT

### CODE REVIEW

**Comment by mgaido91** [comment]
@tdas @zsxwing may I kindly ask if you might take a look at this? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetricsReporter.scala [hunk 1]
```diff
@@ -39,14 +35,18 @@ class MetricsReporter(
 
   // Metric names should not have . in them, so that all the metrics of a query are identified
   // together in Ganglia as a single metric group
-  registerGauge("inputRate-total", () => stream.lastProgress.inputRowsPerSecond)
-  registerGauge("processingRate-total", () => stream.lastProgress.processedRowsPerSecond)
-  registerGauge("latency", () => stream.lastProgress.durationMs.get("triggerExecution").longValue())
-
-  private def registerGauge[T](name: String, f: () => T)(implicit num: Numeric[T]): Unit = {
+  registerGauge("inputRate-total", (lastProgress) => lastProgress.inputRowsPerSecond, 0.0)
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetricsReporter.scala]: nit: `registerGauge("inputRate-total", _.inputRowsPerSecond, 0.0)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetricsReporter.scala [hunk 2]
```diff
@@ -39,14 +35,18 @@ class MetricsReporter(
...
+  registerGauge("processingRate-total", (lastProgress) => lastProgress.processedRowsPerSecond, 0.0)
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetricsReporter.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetricsReporter.scala [hunk 3]
```diff
@@ -39,14 +35,18 @@ class MetricsReporter(
...
+  registerGauge("latency",
+    (lastProgress) => lastProgress.durationMs.get("triggerExecution").longValue(), 0L)
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetricsReporter.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetricsReporter.scala [hunk 4]
```diff
@@ -39,14 +35,18 @@ class MetricsReporter(
...
+
+  private def registerGauge[T](
+      name: String,
+      f: (StreamingQueryProgress) => T,
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetricsReporter.scala]: nit: `(StreamingQueryProgress) => T` -> `StreamingQueryProgress => T,`

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 5]
```diff
@@ -424,6 +424,31 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter with Logging wi
     }
   }
 
+  test("SPARK-22975: MetricsReporter defaults when there was no progress reported") {
+    withSQLConf("spark.sql.streaming.metricsEnabled" -> "true") {
+      BlockingSource.latch = new CountDownLatch(1)
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: You can use `memory` source and `console` sink to simplify the codes like this:
```
  test("SPARK-22975: MetricsReporter defaults when there was no progress reported") {
    withSQLConf("spark.sql.streaming.metricsEnabled" -> "true") {
      withTempDir { tempDir =>
        val sq = MemoryStream[Int].toDF
          .writeStream
          .format("console")
          .start()
          .asInstanceOf[StreamingQueryWrapper]
          .streamingQuery

        val gauges = sq.streamMetrics.metricRegistry.getGauges
        assert(gauges.get("latency").getValue.asInstanceOf[Long] == 0)
        assert(gauges.get("processingRate-total").getValue.asInstanceOf[Double] == 0.0)
        assert(gauges.get("inputRate-total").getValue.asInstanceOf[Double] == 0.0)
        sq.stop()
      }
    }
  }
```

**Review by zsxwing — CHANGES_REQUESTED** [review]
Just some style issues. Otherwise LGTM. Thanks for fixing this!

**Comment by zsxwing** [comment]
Could you update the PR title to include the component such as `[SPARK-22975][SS] MetricsReporter should not throw exception when there was no progress reported`?

**Comment by mgaido91** [comment]
@zsxwing the test you suggested fails on Jenkins while locally it passes. I am no sure about the reason since I cannot reproduce it. I had the same test error in my previous trials. Previously the problem was that running other test cases before this one produced the issue. May you please advice what to do? Can I revert to my previous test?

**Comment by zsxwing** [comment]
@mgaido91 Ah, sorry. There is a race condition in this test since the query may make progress when we are checking the results. Please revert to your previous test. Thanks!

**Comment by zsxwing** [comment]
LGTM pending tests

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
Thanks! Merging to master, 2.3 and 2.2.

**Pull request closed without merging**